File: shortcut_analyzer.cs
// shortcut_analyzer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Shortcut
{
    public string name { get; set; }
    public string shortcut { get; set; }
    public string app { get; set; }
    public string category { get; set; }
}

class Data
{
    public List<Shortcut> shortcuts { get; set; }
}

class Analyzer
{
    private List<Shortcut> shortcuts;
    private Dictionary<string, List<Shortcut>> apps = new();
    private Dictionary<string, List<Shortcut>> categories = new();
    private Dictionary<string, List<Shortcut>> conflicts = new();

    public Analyzer(Data data)
    {
        shortcuts = data.shortcuts ?? new List<Shortcut>();
        Index();
    }

    private void Index()
    {
        foreach (var s in shortcuts)
        {
            string app = string.IsNullOrEmpty(s.app) ? "Без приложения" : s.app;
            if (!apps.ContainsKey(app)) apps[app] = new List<Shortcut>();
            apps[app].Add(s);

            string cat = string.IsNullOrEmpty(s.category) ? "Без категории" : s.category;
            if (!categories.ContainsKey(cat)) categories[cat] = new List<Shortcut>();
            categories[cat].Add(s);

            if (!string.IsNullOrEmpty(s.shortcut))
            {
                if (!conflicts.ContainsKey(s.shortcut)) conflicts[s.shortcut] = new List<Shortcut>();
                conflicts[s.shortcut].Add(s);
            }
        }
    }

    public List<Shortcut> Search(string query)
    {
        string q = query.ToLower();
        return shortcuts.Where(s =>
            s.name.ToLower().Contains(q) ||
            s.shortcut.ToLower().Contains(q)
        ).ToList();
    }

    public Dictionary<string, List<Shortcut>> GetConflicts()
    {
        var result = new Dictionary<string, List<Shortcut>>();
        foreach (var kv in conflicts)
        {
            if (kv.Value.Count > 1) result[kv.Key] = kv.Value;
        }
        return result;
    }

    public Dictionary<string, List<Shortcut>> GroupByApp() => apps;

    public void ExportCSV(string filename)
    {
        using var writer = new StreamWriter(filename);
        writer.WriteLine("name,shortcut,app,category");
        foreach (var s in shortcuts)
            writer.WriteLine($"{s.name},{s.shortcut},{s.app},{s.category}");
        Console.WriteLine(Colorize($"Экспортировано в {filename}", "green"));
    }

    public void ExportJSON(string filename)
    {
        var data = new Data { shortcuts = shortcuts };
        var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine(Colorize($"Экспортировано в {filename}", "green"));
    }

    public void Display(List<Shortcut> items, bool group, bool conflictsOnly, bool verbose)
    {
        var data = items ?? shortcuts;
        if (data.Count == 0)
        {
            Console.WriteLine(Colorize("Нет данных для отображения.", "yellow"));
            return;
        }
        if (conflictsOnly)
        {
            var conf = GetConflicts();
            if (conf.Count == 0)
            {
                Console.WriteLine(Colorize("Конфликтов не найдено.", "green"));
                return;
            }
            Console.WriteLine(Colorize("🔍 Конфликты (одинаковые комбинации):", "bold"));
            foreach (var kv in conf)
            {
                Console.WriteLine(Colorize($"  {kv.Key}:", "red"));
                foreach (var s in kv.Value)
                    Console.WriteLine($"    - {s.name} ({s.app})");
            }
            return;
        }
        if (group)
        {
            var grouped = GroupByApp();
            foreach (var kv in grouped)
            {
                Console.WriteLine(Colorize($"\n📁 {kv.Key} ({kv.Value.Count}):", "blue"));
                foreach (var s in kv.Value)
                    PrintItem(s, verbose);
            }
        }
        else
        {
            foreach (var s in data)
                PrintItem(s, verbose);
        }
    }

    private void PrintItem(Shortcut s, bool verbose)
    {
        if (verbose)
            Console.WriteLine($"  {Colorize(s.name, "bold")} → {Colorize(s.shortcut, "cyan")}  (приложение: {s.app}, категория: {s.category})");
        else
            Console.WriteLine($"  {s.name} → {s.shortcut}");
    }

    private static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "cyan" => "\x1b[96m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static void Main(string[] args)
    {
        string file = null, search = null, exportFile = null;
        bool group = false, conflictsOnly = false, verbose = false;

        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "-s" && i+1 < args.Length) search = args[++i];
            else if (arg == "-g") group = true;
            else if (arg == "-c") conflictsOnly = true;
            else if (arg == "-e" && i+1 < args.Length) exportFile = args[++i];
            else if (arg == "-v") verbose = true;
            else if (arg == "-h" || arg == "--help")
            {
                Console.WriteLine("Usage: shortcut_analyzer <file> [-s search] [-g] [-c] [-e file] [-v]");
                return;
            }
            else if (file == null) file = arg;
        }
        if (file == null)
        {
            Console.WriteLine(Colorize("Укажите файл конфигурации.", "red"));
            return;
        }

        Data data;
        try
        {
            string json = File.ReadAllText(file);
            data = JsonSerializer.Deserialize<Data>(json);
        }
        catch (Exception e)
        {
            Console.WriteLine(Colorize($"Ошибка загрузки файла: {e.Message}", "red"));
            return;
        }

        var analyzer = new Analyzer(data);

        if (search != null)
        {
            var results = analyzer.Search(search);
            if (results.Count > 0)
            {
                Console.WriteLine(Colorize($"🔍 Найдено {results.Count} совпадений:", "bold"));
                analyzer.Display(results, group, false, verbose);
            }
            else Console.WriteLine(Colorize("Ничего не найдено.", "yellow"));
            return;
        }

        if (conflictsOnly)
        {
            analyzer.Display(null, false, true, false);
            return;
        }

        if (exportFile != null)
        {
            if (exportFile.EndsWith(".csv")) analyzer.ExportCSV(exportFile);
            else if (exportFile.EndsWith(".json")) analyzer.ExportJSON(exportFile);
            else Console.WriteLine(Colorize("Неизвестный формат экспорта. Используйте .csv или .json", "red"));
            return;
        }

        analyzer.Display(null, group, false, verbose);
    }
}

File: shortcut_analyzer.java
// shortcut_analyzer.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;
import com.google.gson.*; // install gson

public class shortcut_analyzer {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String CYAN = "\u001B[96m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    static class Shortcut {
        String name;
        String shortcut;
        String app;
        String category;
    }

    static class Data {
        List<Shortcut> shortcuts = new ArrayList<>();
    }

    static class Analyzer {
        List<Shortcut> shortcuts;
        Map<String, List<Shortcut>> apps = new HashMap<>();
        Map<String, List<Shortcut>> categories = new HashMap<>();
        Map<String, List<Shortcut>> conflicts = new HashMap<>();

        Analyzer(Data data) {
            shortcuts = data.shortcuts;
            index();
        }

        void index() {
            for (Shortcut s : shortcuts) {
                String app = s.app == null || s.app.isEmpty() ? "Без приложения" : s.app;
                apps.computeIfAbsent(app, k -> new ArrayList<>()).add(s);

                String cat = s.category == null || s.category.isEmpty() ? "Без категории" : s.category;
                categories.computeIfAbsent(cat, k -> new ArrayList<>()).add(s);

                if (s.shortcut != null && !s.shortcut.isEmpty()) {
                    conflicts.computeIfAbsent(s.shortcut, k -> new ArrayList<>()).add(s);
                }
            }
        }

        List<Shortcut> search(String query) {
            String q = query.toLowerCase();
            return shortcuts.stream()
                .filter(s -> s.name.toLowerCase().contains(q) || s.shortcut.toLowerCase().contains(q))
                .collect(Collectors.toList());
        }

        Map<String, List<Shortcut>> getConflicts() {
            Map<String, List<Shortcut>> result = new HashMap<>();
            for (Map.Entry<String, List<Shortcut>> e : conflicts.entrySet()) {
                if (e.getValue().size() > 1) result.put(e.getKey(), e.getValue());
            }
            return result;
        }

        Map<String, List<Shortcut>> groupByApp() {
            return apps;
        }

        void exportCSV(String filename) throws IOException {
            try (PrintWriter pw = new PrintWriter(filename)) {
                pw.println("name,shortcut,app,category");
                for (Shortcut s : shortcuts) {
                    pw.printf("%s,%s,%s,%s\n", s.name, s.shortcut, s.app, s.category);
                }
            }
            System.out.println(colorize("Экспортировано в " + filename, GREEN));
        }

        void exportJSON(String filename) throws IOException {
            Gson gson = new GsonBuilder().setPrettyPrinting().create();
            Map<String, Object> root = new HashMap<>();
            root.put("shortcuts", shortcuts);
            String json = gson.toJson(root);
            Files.write(Paths.get(filename), json.getBytes());
            System.out.println(colorize("Экспортировано в " + filename, GREEN));
        }

        void display(List<Shortcut> items, boolean group, boolean conflictsOnly, boolean verbose) {
            List<Shortcut> data = items != null ? items : shortcuts;
            if (data.isEmpty()) {
                System.out.println(colorize("Нет данных для отображения.", YELLOW));
                return;
            }
            if (conflictsOnly) {
                Map<String, List<Shortcut>> conf = getConflicts();
                if (conf.isEmpty()) {
                    System.out.println(colorize("Конфликтов не найдено.", GREEN));
                    return;
                }
                System.out.println(colorize("🔍 Конфликты (одинаковые комбинации):", BOLD));
                for (Map.Entry<String, List<Shortcut>> e : conf.entrySet()) {
                    System.out.println(colorize("  " + e.getKey() + ":", RED));
                    for (Shortcut s : e.getValue()) {
                        System.out.println("    - " + s.name + " (" + s.app + ")");
                    }
                }
                return;
            }
            if (group) {
                Map<String, List<Shortcut>> grouped = groupByApp();
                for (Map.Entry<String, List<Shortcut>> e : grouped.entrySet()) {
                    System.out.println(colorize("\n📁 " + e.getKey() + " (" + e.getValue().size() + "):", BLUE));
                    for (Shortcut s : e.getValue()) {
                        printItem(s, verbose);
                    }
                }
            } else {
                for (Shortcut s : data) {
                    printItem(s, verbose);
                }
            }
        }

        void printItem(Shortcut s, boolean verbose) {
            if (verbose) {
                System.out.printf("  %s → %s  (приложение: %s, категория: %s)\n",
                    colorize(s.name, BOLD),
                    colorize(s.shortcut, CYAN),
                    s.app, s.category);
            } else {
                System.out.println("  " + s.name + " → " + s.shortcut);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        String file = null, search = null, exportFile = null;
        boolean group = false, conflictsOnly = false, verbose = false;

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if (arg.equals("-s") && i+1 < args.length) search = args[++i];
            else if (arg.equals("-g")) group = true;
            else if (arg.equals("-c")) conflictsOnly = true;
            else if (arg.equals("-e") && i+1 < args.length) exportFile = args[++i];
            else if (arg.equals("-v")) verbose = true;
            else if (arg.equals("-h") || arg.equals("--help")) {
                System.out.println("Usage: java shortcut_analyzer <file> [-s search] [-g] [-c] [-e file] [-v]");
                return;
            } else if (file == null) file = arg;
        }
        if (file == null) {
            System.out.println(colorize("Укажите файл конфигурации.", RED));
            return;
        }

        Gson gson = new Gson();
        Data data;
        try {
            String json = new String(Files.readAllBytes(Paths.get(file)));
            data = gson.fromJson(json, Data.class);
        } catch (Exception e) {
            System.out.println(colorize("Ошибка загрузки файла: " + e.getMessage(), RED));
            return;
        }

        Analyzer analyzer = new Analyzer(data);

        if (search != null) {
            List<Shortcut> results = analyzer.search(search);
            if (!results.isEmpty()) {
                System.out.println(colorize("🔍 Найдено " + results.size() + " совпадений:", BOLD));
                analyzer.display(results, group, false, verbose);
            } else {
                System.out.println(colorize("Ничего не найдено.", YELLOW));
            }
            return;
        }

        if (conflictsOnly) {
            analyzer.display(null, false, true, false);
            return;
        }

        if (exportFile != null) {
            if (exportFile.endsWith(".csv")) analyzer.exportCSV(exportFile);
            else if (exportFile.endsWith(".json")) analyzer.exportJSON(exportFile);
            else System.out.println(colorize("Неизвестный формат экспорта. Используйте .csv или .json", RED));
            return;
        }

        analyzer.display(null, group, false, verbose);
    }
}
